File: src/com/jetbrains/idear/VoiceRecordControllerAction.java
package com.jetbrains.idear;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.components.ServiceManager;

public class VoiceRecordControllerAction extends AnAction {
    private volatile boolean isRecording = false;

    @Override
    public void actionPerformed(AnActionEvent anActionEvent) {
        ASRService recognitionService = ServiceManager.getService(ASRService.class);
        if (isRecording) {
            isRecording = false;
            getTemplatePresentation().setIcon(Icons.RECORD_START);
            setDefaultIcon(false);
            recognitionService.deactivate();
        }
        else {
            isRecording = true;
            getTemplatePresentation().setIcon(Icons.RECORD_END);
            setDefaultIcon(false);
            recognitionService.activate();
        }
    }

}

File: src/main/java/org/openasr/idear/VoiceRecordControllerAction.kt
package org.openasr.idear

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.components.ServiceManager
import org.acejump.config.AceConfig
import org.acejump.config.AceSettings
import org.openasr.idear.actions.IdearAction
import org.openasr.idear.asr.ASRService
import org.openasr.idear.presentation.Icons

object VoiceRecordControllerAction : IdearAction() {
    @Volatile private var isRecording = false

    override fun actionPerformed(anActionEvent: AnActionEvent) {
      val settings = ServiceManager.getService(AceConfig::class.java).state
      val aceJumpDefaults =settings.allowedChars
        if (isRecording) {
            isRecording = false
            templatePresentation.icon = Icons.RECORD_START
            isDefaultIcon = false
            settings.allowedChars = aceJumpDefaults
            ASRService.deactivate()
        } else {
            isRecording = true
            templatePresentation.icon = Icons.RECORD_END
            isDefaultIcon = false
            settings.allowedChars = "1234567890"
            ASRService.activate()
        }
    }
}